govcd/vm_test.go

```diff
@@ -880,3 +880,46 @@ func (vcd *TestVCD) Test_VMPowerOnPowerOff(check *C) {
 	check.Assert(err, IsNil)
 	check.Assert(vmStatus, Equals, "POWERED_OFF")
 }
+
+// Test gathering VM virtual hardware items
+func (vcd *TestVCD) Test_GetVirtualHardwareSection(check *C) {
+	itemName := "TestGetVirtualHardwareSection"
+
+	if vcd.skipVappTests {
+		check.Skip("Skipping test because vapp wasn't properly created")
+	}
+
+	fmt.Printf("Running: %s\n", itemName)
+
+	// Find VM
+	vapp := vcd.findFirstVapp()
+	vmType, vmName := vcd.findFirstVm(vapp)
+	if vmName == "" {
+		check.Skip("skipping test because no VM is found")
+	}
+	vm, err := vcd.client.Client.FindVMByHREF(vmType.HREF)
+	check.Assert(err, IsNil)
+
+	// Set VirtualHardwareSection of found vm
+	vm.VM.VirtualHardwareSection = &types.VirtualHardwareSection{
+		Item: []*types.VirtualHardwareItem{
+			&types.VirtualHardwareItem{
+				InstanceID:          1,
+				AutomaticAllocation: true,
+				Address:             "1.1.1.1",
+				CoresPerSocket:      2,
+			},
+		},
+	}
+
+	// Preform check of virtual hardware section
+	section, err := vm.GetVirtualHardwareSection()
+	check.Assert(err, IsNil)
+	for _, item := range section.Item {
```

Comment:
The `section.Item` can contain many types of information (see [example output](https://gist.github.com/Didainius/6a01c9154cfc582c36370efde18d0baf)).
You will spot many types of `types.VirtualHardwareItem` inside `types.VirtualHardwareSection` so before checking for specific fields you must be sure that this is the correct item you're looking in.
For example - if you're looking for a MAC address - you must be sure that you're checking it in the `types.VirtualHardwareItem` of network card as checking it in `disk` will not have a MAC set.

_Hint_ the following fields should help to identify type of `types.VirtualHardwareItem`: 
```
<rasd:ResourceSubType>VMXNET3</rasd:ResourceSubType>
<rasd:ResourceType>10</rasd:ResourceType>
```

Reply:
No worries ive actually been out of the country for a while and just getting back up to speed on things. Ahh that example output is exactly what I was looking for before creating this test cause I was not sure what the defaults for these would be. Im wondering though for this check should I get super granular and check all the items or would just checking the top level items, i.e. `<ovf:System>`, `<ovf:Item>`, etc... then maybe also check for the existence of child items under those but not trying to check the specifics of the children?

Reply:
Probably no need to be too granular, but you could check for some value inside top level item (to be sure it is not skipped during un-marshalling)

Reply:
Looking at the types it seems that the only field that would always be there would be `VirtualHardwareSection.Info`, as all the other values in both `VirtualHardwareSection` and `VirtualHardwareItem` are set to `omitempty`. Should one assume that those values would always be there after un-marshelling for this test? Or just simply add the check for `check.Assert(section.Info, NotNil)` knowing that they may or may not be present.

Reply:
You could check that `Info` always exists and also loop through the components and find a specific device based on some field value (ensuring that it is set). Like the network card would have types and subtypes as I mentioned:
```
<rasd:ResourceSubType>VMXNET3</rasd:ResourceSubType>
<rasd:ResourceType>10</rasd:ResourceType>
```

Reply:
Ok so assuming those my latest commit should cover this, however the check within the loop is only validating the `address` field is valid _if_ its a `VMXNET3` network adapter which im assuming will always map to a `ResourceType` of `10`. Should the check also account for other possible types of network adapters or is `ResourceType==10` the only value that _any_ network adapter would have?

Reply:
That is question for research. As far as I have seen the `ResourceType == 10` hints of a network adapter, while `ResourceSubType` hints it's type (VMXNET3, E1000). Should be safe to leave `ResourceType == 10` and then look for a MAC address.